Story/quad 137

PawanMahalle opened the pull request:
@jdamerow @kavinya @rmanojcit  Added functionality to add project blog entries to database. Please review.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogDAO.java:
```diff
@@ -0,0 +1,69 @@
+import edu.asu.spring.quadriga.exceptions.QuadrigaStorageException;
+
+@Service
+public class ProjectBlogDAO extends BaseDAO<ProjectBlogDTO>implements IProjectBlogDAO {
```
Javadoc missing and there needs to be a blank before "implements"

PawanMahalle replied:
Added Javadoc in all DAOs and DTOs added.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogDAO.java:
```diff
@@ -0,0 +1,69 @@
+    public String getCorrespondingProjectID(String projectBlogId)
+            throws QuadrigaStorageException {
+       
+        return null;
```
I'd throw a NotYetImplemented exception here (you would need to create it) or UnsupportedOperation (should exists as a core Java class).

jdamerow replied:
If you create your own `NotYetImplementedException` (would be the preferred way), then you should inherit from `RuntimeException` so that you don't have to declare it in the method signature. This way, once it's implemented you don't have to change all the classes that might use this method. Makes sense?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogDAO.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public void addProjectBlogDTO(ProjectBlogDTO projectBlogDTO) throws QuadrigaStorageException  {
+        
+        projectBlogDTO.setProjectDTO(projectDAO.getProjectDTO(projectBlogDTO.getProjectid()));
```
I wouldn't nest that many method calls. One usually is ok to read, but in this case you nest three levels. Why not have one line that gets the DTO and one line to set it? Or one line to get the Project Id, and one to get and then set the project? This because pretty unwieldy and difficult to read.

jdamerow replied:
Also, looking at it again. I would do this logic on the manager layer, not DB layer. Move this up to the manager and let the DAO only be worried about saving.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dto/ProjectBlogDTO.java:
```diff
@@ -0,0 +1,139 @@
+    @Column(name = "author")
+    private String author;
+    
+    @Lob
```
Why do you use this annotation here?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dto/ProjectBlogDTO.java:
```diff
@@ -0,0 +1,139 @@
+    private String author;
+    
+    @Lob
+    @Basic(optional = false)
```
don't put this kind of restrictions on the database; limiting the code on this level makes it difficult to change later. This kind of logic should be handled in manager and validator classes.

jdamerow replied:
same for all the other optional=false annotations.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/mapper/ProjectBlogDTOMapper.java:
```diff
@@ -0,0 +1,23 @@
+    
+	public ProjectBlogDTO getProjectBlogDTO(IProjectBlog projectBlog) {
+	    
+	    ProjectBlogDTO projectBlogDTO = new ProjectBlogDTO();
```
what's up with the formatting?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+     *             This exception is thrown when user other than project admin
+     *             and owner tries to add project blog entry.
+     */
+    @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 1, userRole = {
```
the add functionality should be in its own controller (roughly remember one controller per page).

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+
+        model.addAttribute("projectBlog", new ProjectBlog());
+        model.addAttribute("project", project);
+        model.addAttribute("project_id", projectId);
```
if you put the project in the model, you don't need to put the id as well, do you? You can just get the id from the project when you need it.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+    @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 3, userRole = {
+            RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
+    @RequestMapping(value = "sites/{projectId}/addprojectblog", method = RequestMethod.POST)
+    public String addprojectblogEntry(@Validated @ModelAttribute("projectBlog") ProjectBlog projectBlog,
```
camel case missing

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+
+        ProjectBlogDTO projectBlogDTO = projectBlogDTOMapper.getProjectBlogDTO(projectBlog);
+        
+        projectBlogDao.addProjectBlogDTO(projectBlogDTO);
```
you are missing a manager class. you controllers should talk to a manager how talks to the DAO. I would also let the manager set the creation date and maybe author, so that if there is a second way of creating blog entries, this code doesn't have to be repeated.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+        
+        projectBlogDao.addProjectBlogDTO(projectBlogDTO);
+        
+        model.addAttribute("project_id", projectId);
```
you need to add the project to the model or you'll break the navigation.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,153 @@
+
+        model.addAttribute("blockentrylist", dummyList);
+        model.addAttribute("project", project);
+        model.addAttribute("project_id", project.getProjectId());
```
Now, that you have the functionality to add entries, you should use the real data here.

jdamerow commented:
see comments; please reopen PR when you're done.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/domain/impl/projectblog/ProjectBlog.java:
```diff
@@ -0,0 +1,180 @@
+ * @author Pawan Mahalle
+ *
+ */
+public class ProjectBlog implements IProjectBlog {
```
This is for one project blog entry, right? Not a whole blog? Then your class should be named accordingly.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/dto/ProjectBlogDTO.java:
```diff
@@ -0,0 +1,130 @@
+@NamedQueries({ 
+    @NamedQuery(name = "ProjectBlogDTO.findAll", query = "SELECT pb FROM ProjectBlogDTO pb"),
+    @NamedQuery(name = "ProjectBlogDTO.findByProjectBlogId", query = "SELECT p FROM ProjectBlogDTO p WHERE p.projectid = :projectId order by createdDate desc"), })
+public class ProjectBlogDTO implements Serializable {
```
Same here: ProjectBlogEntryDTO?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/mapper/ProjectBlogDTOMapper.java:
```diff
@@ -0,0 +1,85 @@
+     * @param projectBlogDTOList
+     * @return
+     */
+    public List<IProjectBlog> getProjectBlogListFromDTOList(List<ProjectBlogDTO> projectBlogDTOList) {
```
I would put this logic in your service class and remove this method.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/projectblog/ProjectBlogManager.java:
```diff
@@ -0,0 +1,64 @@
+    private ProjectBlogDTOMapper projectBlogDTOMapper;
+
+    @Autowired
+    IProjectBlogMapper projectBlogMapper;
```
private

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/projectblog/ProjectBlogManager.java:
```diff
@@ -0,0 +1,64 @@
+
+        List<IProjectBlog> projectBlogList = projectBlogMapper.getProjectBlogListForProject(projectId);
+
+        return projectBlogList;
```
Do the logic you are currently doing in the mapper here.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/projectblog/IProjectBlogManager.java:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * adds new project blog object to database table
+     * <code>tbl_projectblog</code>
```
This is the place to mention that the method will also generate an id and set a creation dat.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/projectblog/mapper/IProjectBlogMapper.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public interface IProjectBlogMapper {
+
+	public List<IProjectBlog> getProjectBlogListForProject(String projectId)
```
Why does your mapper interface have only one method in it? Shouldn't there be 3?

jdamerow replied:
ok, now I'm really confused. What is this class doing?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/projectblog/ProjectBlogManager.java:
```diff
@@ -0,0 +1,64 @@
+    private IProjectBlogDAO projectBlogDao;
+
+    @Autowired
+    private ProjectBlogDTOMapper projectBlogDTOMapper;
```
Autowire the interface here.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/projectblog/mapper/impl/ProjecBlogMapper.java:
```diff
@@ -0,0 +1,40 @@
+    IProjectBlogDAO projectBlogDAO;
+    
+    @Autowired
+    ProjectBlogDTOMapper projectBlogDTOMapper;
```
variables should be private! we talked about that in class remember?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/service/projectblog/mapper/impl/ProjecBlogMapper.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ */
+@Service
+public class ProjecBlogMapper implements IProjectBlogMapper {
```
what is this class for?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogController.java:
```diff
@@ -0,0 +1,102 @@
+ */
+@PropertySource(value = "classpath:/user.properties")
+@Controller
+@Transactional(rollbackFor = { Exception.class })
```
why do you make your controller transactional?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogController.java:
```diff
@@ -0,0 +1,102 @@
+    IUserManager userManager;
+   
+    @Autowired
+    IProjectBlogManager projectBlogManager;
```
should all be private

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogController.java:
```diff
@@ -0,0 +1,102 @@
+    @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 1, userRole = {
+            RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
+    @RequestMapping(value = "sites/{projectId}/addprojectblog", method = RequestMethod.GET)
+    public String addProjectBlogForm(@PathVariable("projectId") String projectId, Model model)
```
Do you mean "projectUnixName" by "projectid"? In the public sites, you should have the project id in the path. Also, merge develop into your branch and check out this:
https://diging.atlassian.net/wiki/display/QUAD/Injecting+projects+by+unix+names

jdamerow replied:
Ah, now I see why you do this. You need to find a different way to achieve what you're doing here. I think for now, I would add the project id as request parameter instead of having it in the path.

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,69 @@
+ */
+@PropertySource(value = "classpath:/user.properties")
+@Controller
+@Transactional(rollbackFor = { Exception.class })
```
same here: why is this transactional?

jdamerow commented on Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java:
```diff
@@ -0,0 +1,69 @@
+     * @throws QuadrigaStorageException
+     */
+    @RequestMapping(value = "sites/{projectUnixName}/projectblog", method = RequestMethod.GET)
+    public String projectblog(@PathVariable("projectUnixName") String projectUnixName, Model model)
```
check out this:
https://diging.atlassian.net/wiki/display/QUAD/Injecting+projects+by+unix+names

jdamerow replied:
Then you can loose line 52.

jdamerow commented on Quadriga/src/main/webapp/WEB-INF/tiles-defs.xml:
```diff
@@ -803,6 +803,16 @@
+	</definition>
+	
+	<definition name="sites/addprojectblog" extends="base.definition.public">
+		<put-attribute name="currentPage" value="addprojectblog" />
```
what entry do you want to have active in the navigation bar? The value for this navigation bar entry should be here (check navbar.jsp).

jdamerow commented:
please reopen when done

PawanMahalle commented:
Changed the following changes in this commit
1. Implemented review changes. Discarded mapper design pattern for data access.
2. Replaced the references to _ProjectBlog_ to _ProjectBlogEntry_.
3. Added @Inject annotation.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
please resolve conflicts.
